### PR TITLE
Feature global pointers

### PR DESCRIPTION
**Pull request by roigcarlo**
This PR add global pointers. They will be used in a near future

### CODE REVIEW

File: kratos/includes/global_pointer.h
Comment (RiccardoRossi) [inline kratos/includes/global_pointer.h]: this makes impossible to use in the same executable MPI and OpenMP alone (without MPI).

i would love to keep that possibility... any way to prserve that?
to explain myself more precisely, the scenario i see is the following:

i am on a cluster, and i compile with mpi=0n. Now i want to run OpenMP alone. I will still need mpirun...while this is definitely not what i would want. An excellent example of this is to run the tests

File: kratos/includes/global_pointer.h
Comment (RiccardoRossi) [inline kratos/includes/global_pointer.h]: same comment as above...

**Comment by philbucher** [comment]
@roigcarlo are you planning on implementing a parallel data exchange for this?

**Comment by roigcarlo** [comment]
@philbucher Yes and no. The final goal is actually a search algorithm in which almost all the data is kept in the owner process. Think of it a pseudo-mapreduceish kind of scheme.

At some point you will have a container of global pointers in which you have a function to apply and a "worker" that will distribute the work based on where the pointers are located.

Veeery schematically:

```C++
void DoOperation(Data) {
   ....
}

Apply(vector<GlobalPointer> Data, Function=DoOperation) {
    DataProc = GroupByProcID(Data) 

    for(every proc) {
        if local_proc
           DoOperation(DataProc[thisProc])
        else
           SendData(DataProc[thisProc])
    }
}
```

So you will basically have a pointer in your process which points to data in other process, but you will be able to operate normally via the worker and perform simple operations with it ( count, move, etc...)

File: kratos/includes/global_pointer.h
Comment (pooyan-dadvand) [inline kratos/includes/global_pointer.h]: I believe that we should do the same as boost::share_ptr

File: kratos/includes/global_pointer.h
Comment (pooyan-dadvand) [inline kratos/includes/global_pointer.h]: Is there any specific reason to not create it via a weak pointer?

File: kratos/includes/global_pointer.h
Comment (pooyan-dadvand) [inline kratos/includes/global_pointer.h]: Would you please add the copy constructor, move constructor and copy and move = operator?

File: kratos/tests/global_pointers/test_global_pointer.cpp
Comment (pooyan-dadvand) [inline kratos/tests/global_pointers/test_global_pointer.cpp]: local variables should be like form_raw

File: kratos/tests/global_pointers/test_global_pointer.cpp
Comment (pooyan-dadvand) [inline kratos/tests/global_pointers/test_global_pointer.cpp]: I think we should add a macro for MPI test cases which deos the enable disable. But it would be subject of another branch. It is ok as you did now.

File: kratos/tests/global_pointers/test_global_pointer.cpp
Comment (pooyan-dadvand) [inline kratos/tests/global_pointers/test_global_pointer.cpp]: It would be better to use new and delete

**Review by pooyan-dadvand — COMMENTED** [review]
It is ok and aligend with our design! good job!

Nevertheless there are some minor improvements

**Comment by philbucher** [comment]
@roigcarlo this is what I meant, will this worker do data exchange btw ranks in the background? In your example the `Apply` function would be the worker that also exchanges the data?
Sorry closed by accident

**Comment by roigcarlo** [comment]
@philbucher  Yep exactly

**Comment by philbucher** [comment]
@roigcarlo and the `Apply` function would be provided or implemented ourselves?

**Comment by roigcarlo** [comment]
Apply function will be provided by us. but I don't want to advance things since we still have to discuss with @pooyan-dadvand implementation details

**Comment by philbucher** [comment]
ok thx for the explanation

File: kratos/includes/global_pointer.h
Comment (roigcarlo) [reply kratos/includes/global_pointer.h]: I suppose no. anyway, its odd since this IMHO can be seen from two different sides:

1 - We use the current implementation of the global_ptr:
It means that the lock needs to be done when acquiring the data, so once the user creates the global_ptr, the old weak_ptr is useless, as one can easily bypass the lock by accessing the data directly from the global_ptr.

2 - We store the weak_ptr inside the global_ptr:
Then the lock can be made when "using" the pointer, but it needs a completely separate implementation in order to ensure that honors the lock mechanism.

I uploaded following 1, but I would not recommend using this...

**Comment by roigcarlo** [comment]
@pooyan-dadvand I have addressed your comments ( I hope all of them ). Just some small comments with the weak_ptr...

Also, about the rank value with single process runs. I am getting rank 0 systematically but I cannot find it in the doc yet.

**Comment by pooyan-dadvand** [comment]
I have not seen the move constructor and assignement. Have you commit it? (Actually the default one would work for this case)

**Comment by roigcarlo** [comment]
Ups your right, sorry. Lost in another branch, I'll upload them now

**Comment by roigcarlo** [comment]
@jcotela @RiccardoRossi @pooyan-dadvand 
There is another detail before merging. 
As you see there are tests for the MPI version. This has two problems:

1 - You cannot enable MPI in the kratos core at will because _CMakeList.txt_:
```C++
#finding Mpi

# VVVVVVVVVVVVVVVVV
set(MPI_NEEDED OFF)
# ^^^^^^^^^^^^^^^^^

if(${METIS_APPLICATION} MATCHES ON )
  set(MPI_NEEDED ON)
endif(${METIS_APPLICATION} MATCHES ON )
# etc...
```

That line prevents you from enabling it in the configure.

2 - Also, once the code compiles with the MPI enabled, as is part of Kratos core it need to link against mpi, which does not.

So, we were discussing with @pooyan-dadvand to either move the mpi test somewhere else, or make kratos link with MPI if MPI_NEEDED=ON.

Thoughts?

**Comment by philbucher** [comment]
I think having a flag in the `configure.sh` for enabling MPI is much clearer. I also thought of suggesting this

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
guys, i block this as i think that this point should be addressed (as we discussed on friday at lunch)
please take a look to my comments in the thread. in particular there is one that says the following:

using a "#ifdef MPI" makes impossible to use in the same executable MPI and OpenMP alone (without MPI).

i would love to keep that possibility... any way to prserve that?
to explain myself more precisely, the scenario i see is the following:

i am on a cluster, and i compile with mpi=0n. Now i want to run OpenMP alone. I will still need mpirun...while this is definitely not what i would want. An excellent example of this is to run the tests

**Comment by roigcarlo** [comment]
@RiccardoRossi Why can't you run the program with MPI?. I was not there during the discussion you had about this.

Is this about the singleton MPI_INIT if compiled with mpich/pmi?

**Comment by jcotela** [comment]
@roigcarlo The problem is as follows: assume I compiled my kratos with MPI support, but now I want to do an OpenMP run. If your object does special things when compiled with MPI, it will try to do them in my runs but, since I didn't call MPI_Init at any point, it will fail. Your only option then is to add MPI_Initialized calls to check for MPI status everywhere, which is very cumbersome.

**Comment by pooyan-dadvand** [comment]
The only way that I see is to have the rank an static member of the Kernel: `Kernel::GetRank()` and having an MPIKernel which initialize it to the mpi one. (we are discussing this with @jcotela  for a while now)

File: kratos/includes/global_pointer.h
Comment (RiccardoRossi) [reply kratos/includes/global_pointer.h]: @pooyan-dadvand @jcotela 

i am ok with having rank in the kernel, however i would call it something like "unique_processor_id" or smthg of the sort.

my point is that one may have different modelparts employing different communicator worlds

Comment (pooyan-dadvand) [reply kratos/includes/global_pointer.h]: @RiccardoRossi the unique process id would mix the MPI and openMP?

**Comment by jcotela** [comment]
@pooyan-dadvand and I were discussing this the other day and we agreed that we need to understand a bit better how to use separate communicators (for example: one mpi comm for the structure and another, non-overlapping one for the fluid) before deciding on the issue of determining the rank.

The main point is *what does the global pointer need the rank for*. Consider a scenario where ranks 0-9 solve the fluid and ranks 10-19 solve the structure. If we use global pointers within the structure, they can use the rank indices for MPI_COMM_WORLD (10-19) or the structure communicator rank indices (0-9).

Using 10-19 is good if we want to have *unique* indices for the pointers, but useless to exchange information using the structure communicator. Conversely, using 0-9 is good for communication within the structure but not unique (so a global pointer with a rank of 9 will refer to different processes depending on the communicator used).

As an aside, since I had to read a bit on these smaller communicators, I'm linking to a small [tutorial](http://mpitutorial.com/tutorials/introduction-to-groups-and-communicators/) I found useful.

File: kratos/includes/global_pointer.h
Comment (pooyan-dadvand) [inline kratos/includes/global_pointer.h]: It mustn't be 

``` c++
TDataType* GetPointer() { return mDataPointer; }
```

File: kratos/includes/global_pointer.h
Comment (pooyan-dadvand) [inline kratos/includes/global_pointer.h:131]: missing the constant version of operator ->

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
Comments are given....

**Review by pooyan-dadvand — DISMISSED** [review]
For me it is ok, Just to comment that the common world is not stored inside which means that the caller should know in which world is. I'm remarking this because in our design with @jcotela we thought that it should store it. 

@jcotela do you have any comment? @RiccardoRossi is this ok for you?

**Review by RiccardoRossi — DISMISSED** [review]

File: kratos/includes/global_pointer.h
Comment (RiccardoRossi) [inline kratos/includes/global_pointer.h]: how will you know when you are within an appliction that needs mpi, that the core was compiled with mpi?

**Comment by pooyan-dadvand** [comment]
@roigcarlo After discussing with @RiccardoRossi we arrived to conclusion that it would be better to remove `KRATOS_MPI` macros and leave global pointer with overhead of rank all the time.

Do you agree?

File: applications/mpi_search_application/custom_python/add_custom_utilities_to_python.cpp
Comment (RiccardoRossi) [inline applications/mpi_search_application/custom_python/add_custom_utilities_to_python.cpp]: minor: i don't think this is a very good name. Maybe "Repartition"? (just ignore the comment if you like the current one)

**Comment by RiccardoRossi** [comment]
thx @pooyan-dadvand on my side that dissipates all of my concerns

**Comment by philbucher** [comment]
hi guys do you have a plan of when to merge this?

**Comment by pooyan-dadvand** [comment]
@roigcarlo I still see the `KRATOS_USING_MPI` macros. Do you agree in removing them?

**Comment by roigcarlo** [comment]
Sure, we speak about removing this already but was never implemented

**Comment by pooyan-dadvand** [comment]
so, would you please proceed with implementation?

**Comment by roigcarlo** [comment]
@pooyan-dadvand Done

File: kratos/includes/global_pointer.h
Comment (philbucher) [inline kratos/includes/global_pointer.h]: `Kratos:: ...`?

Comment (roigcarlo) [reply kratos/includes/global_pointer.h]: Not sure what you mean, something wrong there?

Comment (philbucher) [reply kratos/includes/global_pointer.h]: nothing wrong, was just wondering why you don't use the new `Kratos::shared_ptr` => #1328

Comment (roigcarlo) [reply kratos/includes/global_pointer.h]: :+1:  But I will make the changes in a separate PR, since this is very old and if I merge the master with this it will crash

**Comment by roigcarlo** [comment]
@RiccardoRossi or @pooyan-dadvand is this ok? can we merge this?

File: kratos/includes/global_pointer.h
Comment (pooyan-dadvand) [inline kratos/includes/global_pointer.h]: `save` is lower case in all Kratos.

File: kratos/includes/global_pointer.h
Comment (pooyan-dadvand) [inline kratos/includes/global_pointer.h]: `load` is lower case in all Kratos.

File: kratos/tests/global_pointers/test_global_pointer.cpp
Comment (pooyan-dadvand) [inline kratos/tests/global_pointers/test_global_pointer.cpp]: We should change this to Kratos:: (maybe afterward)

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
More comments...

File: kratos/tests/global_pointers/test_global_pointer.cpp
Comment (roigcarlo) [reply kratos/tests/global_pointers/test_global_pointer.cpp]: Yep. As I said to @philbucher I will do this once its merged

**Comment by roigcarlo** [comment]
@pooyan-dadvand According to @RiccardoRossi this has more priority now. Honestly I did not remember the status of this but seems to be finished, but we cannot merge it because this is prior to pybind etc... I will adapt it so we can merge if you agree

**Comment by loumalouomega** [comment]
@philbucher and I will be very happy of this. With this the contact and mapping implementation will be easier on MPI

**Comment by philbucher** [comment]
I recently started implementing a more general mapper, but without using this.
I will reevaluate if I can use it once it is merged

**Review by RiccardoRossi — COMMENTED** [review]
just to say that looking at this today, with the rank being passed always and being equal to zero in openmp case i am quite ok! It does not make sense to approve it now, since it will simply not compile, however i am positive on this

**Comment by roigcarlo** [comment]
@pooyan-dadvand This should be usable now.

**Review by RiccardoRossi — DISMISSED** [review]
this looks fine to me, apart for not having any test for mpi search (i think at least)

I am approving...however if you could find the time to add a small mpi search test it would be perfect

File: kratos/includes/global_pointer.h
Comment (RiccardoRossi) [reply kratos/includes/global_pointer.h:131]: @pooyan-dadvand this is done, i would say you could dismiss your review

File: applications/mpi_search_application/custom_python/add_custom_utilities_to_python.cpp
Comment (pooyan-dadvand) [reply applications/mpi_search_application/custom_python/add_custom_utilities_to_python.cpp]: Repartition is definitely a better name.

File: kratos/includes/global_pointer.h
Comment (pooyan-dadvand) [reply kratos/includes/global_pointer.h]: Please apply this comment

File: kratos/includes/global_pointer.h
Comment (pooyan-dadvand) [reply kratos/includes/global_pointer.h]: here the same

**Review by pooyan-dadvand — COMMENTED** [review]
There are still some minor naming inconsistencies to be finished.

**Comment by roigcarlo** [comment]
@pooyan-dadvand I've changed the Repartiton one, but I think you were reviewing an old version of the global_pointer.h file....

**Review by pooyan-dadvand — COMMENTED** [review]
And load and save please

**Comment by pooyan-dadvand** [comment]
Now I see, they were changed!

**Review by pooyan-dadvand — APPROVED** [review]

**Pull request merged**
